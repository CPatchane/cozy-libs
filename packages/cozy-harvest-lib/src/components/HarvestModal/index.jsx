import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import Modal, { ModalDescription, ModalHeader } from 'cozy-ui/react/Modal'
import AppIcon from 'cozy-ui/react/AppIcon'
import Spinner from 'cozy-ui/react/Spinner'

import HarvestModalContent from './Content'
import withAccountFromTrigger from '../HOCs/withAccountFromTrigger'

export class HarvestModal extends PureComponent {
  constructor(props) {
    super(props)
    this.fetchIcon = this.fetchIcon.bind(this)
  }

  fetchIcon() {
    const { client } = this.context
    const { konnector } = this.props
    return client.stackClient.getIconURL({
      type: 'konnector',
      slug: konnector.slug
    })
  }

  render() {
    const {
      account,
      dismissAction,
      fetchingAccount,
      konnector,
      into,
      trigger,
      onLoginSuccess,
      onSuccess,
      running
    } = this.props

    return (
      <Modal
        dismissAction={dismissAction}
        mobileFullscreen
        size="small"
        into={into}
      >
        <ModalHeader>
          <AppIcon fetchIcon={this.fetchIcon} className="u-mah-3 u-ml-1" />
        </ModalHeader>
        <ModalDescription>
          {fetchingAccount ? (
            <Spinner
              size="xxlarge"
              className="u-flex u-flex-justify-center u-pv-3"
            />
          ) : (
            <HarvestModalContent
              account={account}
              konnector={konnector}
              showError={false}
              trigger={trigger}
              running={running}
              onLoginSuccess={onLoginSuccess}
              onSuccess={onSuccess}
            />
          )}
        </ModalDescription>
      </Modal>
    )
  }
}

HarvestModal.contextTypes = {
  client: PropTypes.object.isRequired
}

export default withAccountFromTrigger(HarvestModal)
