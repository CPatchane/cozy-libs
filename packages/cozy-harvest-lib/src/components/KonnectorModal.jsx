import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import { withMutations } from 'cozy-client'
import AppIcon from 'cozy-ui/react/AppIcon'
import Modal, { ModalContent, ModalHeader } from 'cozy-ui/react/Modal'
import Spinner from 'cozy-ui/react/Spinner'
import { SubTitle } from 'cozy-ui/react/Text'

import accountMutations from '../connections/accounts'
import * as triggers from '../helpers/triggers'
import TriggerManager from './TriggerManager'
import withLocales from './hoc/withLocales'

/**
 * KonnectorModal open a Modal related to a given konnector. It fetches the
 * first account and then include a TriggerManager component.
 *
 * This component is aimed to offer an UI to manage all the konnector related
 * triggers and accounts.
 */
export class KonnectorModal extends PureComponent {
  state = {
    account: null,
    fetching: true,
    trigger: null
  }

  constructor(props) {
    super(props)
    this.fetchIcon = this.fetchIcon.bind(this)
  }

  componentDidMount() {
    this.fetchAccount()
  }

  async fetchAccount() {
    const { findAccount } = this.props
    this.setState({ fetching: true })
    const { konnector } = this.props
    const trigger = konnector.triggers.data[0]
    const account = await findAccount(triggers.getAccountId(trigger))
    this.setState({
      account,
      fetching: false
    })
  }

  fetchIcon() {
    const { client } = this.context
    const { konnector } = this.props
    return client.stackClient.getIconURL({
      type: 'konnector',
      slug: konnector.slug
    })
  }

  render() {
    const {
      dismissAction,
      konnector,
      into,
      onLoginSuccess,
      onSuccess,
      running,
      t
    } = this.props

    const { account, fetching, trigger } = this.state

    return (
      <Modal
        dismissAction={dismissAction}
        mobileFullscreen
        size="small"
        into={into}
      >
        <ModalHeader>
          <AppIcon fetchIcon={this.fetchIcon} className="u-mah-3 u-ml-1" />
        </ModalHeader>
        {fetching ? (
          <ModalContent>
            <Spinner
              size="xxlarge"
              className="u-flex u-flex-justify-center u-pv-3"
            />
          </ModalContent>
        ) : (
          <ModalContent>
            <SubTitle className="u-mb-1 u-ta-center">
              {t('modal.konnector.title', { name: konnector.name })}
            </SubTitle>
            {error ? (
              <Infos
                actionButton={
                  <Button theme="danger">
                    {t('modal.konnector.error.button')}
                  </Button>
                }
                title={t('modal.konnector.error.title')}
                text={t('modal.konnector.error.description', error)}
                isImportant
              />
            ) : (
              <TriggerManager
                account={account}
                konnector={konnector}
                trigger={trigger}
                running={running}
                onLoginSuccess={onLoginSuccess}
                onSuccess={onSuccess}
              />
            )}
          </ModalContent>
        )}
      </Modal>
    )
  }
}

KonnectorModal.contextTypes = {
  client: PropTypes.object.isRequired
}

export default withMutations(accountMutations)(withLocales(KonnectorModal))
