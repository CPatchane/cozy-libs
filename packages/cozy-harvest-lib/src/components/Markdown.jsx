import React from 'react'
import ReactMarkdown from 'react-markdown'

export const reactMarkdownRendererOptions = {
  // eslint-disable-next-line react/display-name
  Link: props => (
    <a href={props.href} rel="noopener noreferrer" target="_blank">
      {props.children}
    </a>
  ),
  // eslint-disable-next-line react/display-name
  paragraph: props => <p className="u-mv-0">{props.children}</p>
}

export const Markdown = ({ source }) => (
  <ReactMarkdown source={source} renderers={reactMarkdownRendererOptions} />
)

export default Markdown
