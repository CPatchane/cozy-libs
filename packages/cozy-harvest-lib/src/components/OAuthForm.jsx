import React, { PureComponent } from 'react'

import Button from 'cozy-ui/react/Button'
import { translate } from 'cozy-ui/react/I18n'

export class OAuthForm extends PureComponent {
  render() {
    const { oauth, t } = this.props
    return oauth ? null : (
      <Button
        className="u-mt-1"
        extension="full"
        label={t('oauth.connect.label')}
      />
    )
  }
}

export default translate()(OAuthForm)
