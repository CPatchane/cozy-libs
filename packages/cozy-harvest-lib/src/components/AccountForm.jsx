import React, { PureComponent } from 'react'
import { Form, Field as FinalFormField } from 'react-final-form'
import PropTypes from 'react-proptypes'

import Button from 'cozy-ui/react/Button'
import { translate, extend } from 'cozy-ui/react/I18n'
import Field from 'cozy-ui/react/Field'

import Manifest from '../Manifest'
import OAuthForm from './OAuthForm'

const ENCRYPTED_PLACEHOLDER = '*************'

const predefinedLabels = [
  'answer',
  'birthdate',
  'code',
  'date',
  'email',
  'firstname',
  'lastname',
  'login',
  'password',
  'phone'
]

export class AccountField extends PureComponent {
  render() {
    const { label, name, t, type } = this.props

    // Allow manifest to specify predefined label
    const localeKey = predefinedLabels.includes(label) ? label : name

    const fieldProps = {
      ...this.props,
      className: 'u-m-0', // 0 margin
      fullwidth: true,
      label: t(`fields.${localeKey}.label`, { _: name }),
      size: 'medium'
    }
    const passwordLabels = {
      hideLabel: t('accountForm.password.hide'),
      showLabel: t('accountForm.password.show')
    }
    switch (type) {
      case 'dropdown':
        return (
          <Field
            {...fieldProps}
            options={fieldProps.options.map(option => ({
              ...option,
              // legacy
              label: option.name
            }))}
            value={fieldProps.options.find(o => o.value === fieldProps.value)}
            type="select"
          />
        )
      case 'password':
        return <Field {...fieldProps} secondaryLabels={passwordLabels} />
      default:
        return <Field {...fieldProps} type="text" />
    }
  }
}

// As SelectBox component from Cozy-UI, rendering dropdown type, is just giving
// us the full Option object, we just get its value to facilitate mapping
// with account
const parse = type => value => {
  return type === 'dropdown' ? value.value : value
}

export class AccountFields extends PureComponent {
  render() {
    const { fillEncrypted, manifestFields, t } = this.props

    // Ready to use named fields array
    const namedFields = Object.keys(manifestFields).map(fieldName => ({
      ...manifestFields[fieldName],
      name: fieldName
    }))

    return (
      <div>
        {namedFields.map((field, index) => (
          <FinalFormField
            key={index}
            name={field.name}
            parse={parse(field.type)}
          >
            {({ input }) => (
              <AccountField
                {...field}
                {...input}
                placeholder={
                  field.encrypted && fillEncrypted
                    ? ENCRYPTED_PLACEHOLDER
                    : field.placeholder
                }
                t={t}
              />
            )}
          </FinalFormField>
        ))}
      </div>
    )
  }
}

export class AccountForm extends PureComponent {
  constructor(props, context) {
    super(props, context)
    const { lang, locales } = props
    if (locales && lang) {
      extend(locales[lang])
    }
  }

  render() {
    const { account, fields, oauth, t } = this.props

    if (oauth) return <OAuthForm oauth={oauth} />

    const sanitizedFields = Manifest.sanitizeFields(fields)
    const initialValues = account ? account.auth : {}
    return (
      <Form
        initialValues={initialValues}
        // eslint-disable-next-line no-console
        onSubmit={v => console.log(v)}
        render={({ values }) => (
          <div>
            <AccountFields
              fillEncrypted={!!account}
              manifestFields={sanitizedFields}
              t={t}
            />
            <Button
              className="u-mt-2 u-mb-1-half"
              extension="full"
              label={t('accountForm.submit.label')}
              onclick={() => alert(JSON.stringify(values, 0, 2))}
            />
          </div>
        )}
      />
    )
  }
}

AccountForm.propTypes = {
  account: PropTypes.object,
  fields: PropTypes.object.isRequired,
  locales: PropTypes.object
}

export default translate()(AccountForm)
