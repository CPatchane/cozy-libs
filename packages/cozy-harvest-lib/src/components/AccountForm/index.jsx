import React, { PureComponent } from 'react'
import { Form, Field as FinalFormField } from 'react-final-form'
import PropTypes from 'react-proptypes'

import Button from 'cozy-ui/react/Button'
import { translate, extend } from 'cozy-ui/react/I18n'

import AccountField from './AccountField'
import AccountFormError from './Error'
import { getEncryptedFieldName } from '../../helpers/fields'
import { KonnectorJobError } from 'helpers/konnectors'
import Manifest from 'Manifest'
import OAuthForm from 'components/OAuthForm'

const VALIDATION_ERROR_REQUIRED_FIELD = 'VALIDATION_ERROR_REQUIRED_FIELD'

// As SelectBox component from Cozy-UI, rendering dropdown type, is just giving
// us the full Option object, we just get its value to facilitate mapping
// with account
const parse = type => value => {
  return type === 'dropdown' ? value.value : value
}

export class AccountFields extends PureComponent {
  render() {
    const {
      container,
      disabled,
      hasError,
      initialValues,
      manifestFields,
      onKeyUp,
      t
    } = this.props

    // Ready to use named fields array
    const namedFields = Object.keys(manifestFields).map(fieldName => ({
      ...manifestFields[fieldName],
      name: fieldName
    }))

    return (
      <div onKeyUp={onKeyUp}>
        {namedFields.map((field, index) => (
          <FinalFormField
            key={index}
            name={field.name}
            parse={parse(field.type)}
          >
            {({ input }) => (
              <AccountField
                {...field}
                {...input}
                container={container}
                hasError={hasError}
                disabled={disabled}
                initialValue={
                  initialValues[field.name] ||
                  initialValues[getEncryptedFieldName(field.name)]
                }
                t={t}
              />
            )}
          </FinalFormField>
        ))}
      </div>
    )
  }
}

AccountFields.propTypes = {
  disabled: PropTypes.bool,
  hasError: PropTypes.bool,
  fillEncrypted: PropTypes.bool,
  manifestFields: PropTypes.object.isRequired,
  onKeyUp: PropTypes.func,
  t: PropTypes.func
}

export class AccountForm extends PureComponent {
  constructor(props, context) {
    super(props, context)
    const { konnector, lang } = props
    const { locales } = konnector
    if (locales && lang) {
      extend(locales[lang])
    }
  }

  /**
   * Indicates if the state of ReactFinalForm implies that data can be submitted
   * @param  {Object} formState See https://github.com/final-form/final-form#formstate
   * @return {Boolean}
   */
  isSubmittable({ dirty, error, initialValues, valid }) {
    const untouched = initialValues && !dirty
    return error || (valid && !untouched)
  }

  handleKeyUp(event, { dirty, form, initialValues, valid, values }) {
    if (
      event.code === 'Enter' &&
      this.isSubmittable({ dirty, initialValues, valid })
    ) {
      this.handleSubmit(values, form)
    }
  }

  handleSubmit(values, form) {
    const { onSubmit } = this.props
    // Reset form with new values to set back dirty to false
    form.reset(values)
    onSubmit(values)
  }

  validate = (fields, initialValues) => vals => {
    let errors = {}
    for (let name in fields)
      if (
        fields[name].required &&
        !vals[name] &&
        // Don't require value for empty encrypted fields with initial value
        !initialValues[getEncryptedFieldName(name)]
      )
        errors[name] = VALIDATION_ERROR_REQUIRED_FIELD
    return errors
  }

  render() {
    const {
      account,
      error,
      konnector,
      onSubmit,
      showError,
      submitting,
      t
    } = this.props
    const { fields, oauth } = konnector

    if (oauth) return <OAuthForm initialValues={initialValues} oauth={oauth} />

    const sanitizedFields = Manifest.sanitizeFields(fields)
    const defaultValues = Manifest.defaultFieldsValues(sanitizedFields)
    const initialValues = account && account.auth
    const initialAndDefaultValues = { ...defaultValues, ...initialValues }

    let container = null

    const isLoginError =
      error instanceof KonnectorJobError && error.isLoginError()

    return (
      <Form
        initialValues={initialAndDefaultValues}
        onSubmit={onSubmit}
        validate={this.validate(sanitizedFields, initialAndDefaultValues)}
        render={({ dirty, form, values, valid }) => (
          <div
            ref={element => {
              container = element
            }}
          >
            {error && (showError || isLoginError) && (
              <AccountFormError error={error} konnector={konnector} t={t} />
            )}
            <AccountFields
              container={container}
              disabled={submitting}
              hasError={
                error &&
                error instanceof KonnectorJobError &&
                error.isLoginError()
              }
              initialValues={initialAndDefaultValues}
              manifestFields={sanitizedFields}
              onKeyUp={event =>
                this.handleKeyUp(event, {
                  dirty,
                  form,
                  initialValues,
                  valid,
                  values
                })
              }
              t={t}
            />
            <Button
              busy={submitting}
              className="u-mt-2 u-mb-1-half"
              disabled={
                submitting ||
                !this.isSubmittable({ dirty, error, initialValues, valid })
              }
              extension="full"
              label={t('accountForm.submit.label')}
              onClick={() => this.handleSubmit(values, form)}
            />
          </div>
        )}
      />
    )
  }
}

AccountForm.propTypes = {
  account: PropTypes.object,
  konnector: PropTypes.object.isRequired,
  error: PropTypes.object,
  showError: PropTypes.bool,
  submitting: PropTypes.bool
}

AccountForm.defaultProps = {
  showError: true
}

export default translate()(AccountForm)
