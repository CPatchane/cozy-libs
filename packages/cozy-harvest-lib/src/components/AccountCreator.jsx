import React, { PureComponent } from 'react'
import PropTypes from 'react-proptypes'

import { withMutations } from 'cozy-client'

import AccountForm from './AccountForm'
import { accountsMutations } from '../connections/accounts'
import { mergeAuth, prepareAccountData } from '../helpers/accounts'

/**
 * Encapsulates an AccountForm and create an account with resulting data.
 */
export class AccountCreator extends PureComponent {
  constructor(props) {
    super(props)
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  async handleSubmit(data) {
    const {
      createAccount,
      konnector,
      onBeforeCreate,
      onCreateSuccess
    } = this.props
    onBeforeCreate()
    const account = await createAccount(
      konnector,
      prepareAccountData(konnector, data)
    )
    // Merge auth to keep original values for encrypted fields during creation
    onCreateSuccess(mergeAuth(account, data))
  }

  render() {
    const { account, konnector, submitting } = this.props
    return (
      <AccountForm
        fields={konnector.fields}
        initialValues={account && (account.auth || account.oauth)}
        locales={konnector.locales}
        oauth={konnector.oauth}
        onSubmit={this.handleSubmit}
        submitting={submitting}
      />
    )
  }
}

AccountCreator.propTypes = {
  createAccount: PropTypes.func.isRequired,
  konnector: PropTypes.object.isRequired,
  submitting: PropTypes.bool,
  onBeforeCreate: PropTypes.func.isRequired,
  onCreateSuccess: PropTypes.func.isRequired
}

export default withMutations(accountsMutations)(AccountCreator)
