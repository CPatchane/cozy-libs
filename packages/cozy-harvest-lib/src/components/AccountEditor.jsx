import React, { PureComponent } from 'react'
import PropTypes from 'react-proptypes'

import { withMutations } from 'cozy-client'

import AccountForm from './AccountForm'
import { accountsMutations } from '../connections/accounts'
import { mergeAuth } from '../helpers/accounts'

/**
 * Encapsulates an AccountForm of an existing account, and allows to update it.
 */
export class AccountEditor extends PureComponent {
  constructor(props) {
    super(props)
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  async handleSubmit(data) {
    const {
      account,
      onBeforeUpdate,
      onUpdateSuccess,
      updateAccount
    } = this.props
    onBeforeUpdate(account)
    const updatedAccount = await updateAccount(mergeAuth(account, data))
    onUpdateSuccess(updatedAccount)
  }

  render() {
    const { account, konnector, submitting } = this.props
    return (
      <AccountForm
        fields={konnector.fields}
        initialValues={account && (account.auth || account.oauth)}
        locales={konnector.locales}
        oauth={konnector.oauth}
        onSubmit={this.handleSubmit}
        submitting={submitting}
      />
    )
  }
}

AccountEditor.propTypes = {
  account: PropTypes.object.isRequired,
  konnector: PropTypes.object.isRequired,
  onBeforeUpdate: PropTypes.func.isRequired,
  onUpdateSuccess: PropTypes.func.isRequired,
  submitting: PropTypes.bool,
  updateAccount: PropTypes.func.isRequired
}

export default withMutations(accountsMutations)(AccountEditor)
