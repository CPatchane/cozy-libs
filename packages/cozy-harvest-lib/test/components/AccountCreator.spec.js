/* eslint-env jest */
import React from 'react'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import { AccountCreator } from 'components/AccountCreator'

configure({ adapter: new Adapter() })

const onCreateSpy = jest.fn()
const onCreateSuccessSpy = jest.fn()

const createAccountMock = jest.fn()

const fixtures = {
  konnector: {
    slug: 'konnectest'
  },
  data: {
    username: 'foo',
    passphrase: 'bar'
  },
  expectedAccount: {
    account_type: 'konnectest',
    auth: {
      username: 'foo',
      passphrase: 'bar'
    }
  }
}

describe('AccountCreator', () => {
  it('should render', () => {
    const component = shallow(
      <AccountCreator
        createAccount={createAccountMock}
        konnector={fixtures.konnector}
        onBeforeCreate={onCreateSpy}
        onCreateSuccess={onCreateSuccessSpy}
      />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  describe('handleSubmit', () => {
    beforeEach(() => {
      createAccountMock.mockResolvedValue(fixtures.expectedAccount)
    })

    afterEach(() => {
      createAccountMock.mockReset()
      onCreateSpy.mockClear()
      onCreateSuccessSpy.mockClear()
    })

    it('should call onCreate', () => {
      const wrapper = shallow(
        <AccountCreator
          createAccount={createAccountMock}
          konnector={fixtures.konnector}
          onBeforeCreate={onCreateSpy}
          onCreateSuccess={onCreateSuccessSpy}
        />
      )
      wrapper.instance().handleSubmit(fixtures.data)
      expect(onCreateSpy).toHaveBeenCalledTimes(1)
    })

    it('should call onCreateSuccess', async () => {
      const wrapper = shallow(
        <AccountCreator
          createAccount={createAccountMock}
          konnector={fixtures.konnector}
          onBeforeCreate={onCreateSpy}
          onCreateSuccess={onCreateSuccessSpy}
        />
      )
      await wrapper.instance().handleSubmit(fixtures.data)
      expect(onCreateSuccessSpy).toHaveBeenCalledTimes(1)
      expect(onCreateSuccessSpy).toHaveBeenCalledWith(fixtures.expectedAccount)
    })
  })
})
