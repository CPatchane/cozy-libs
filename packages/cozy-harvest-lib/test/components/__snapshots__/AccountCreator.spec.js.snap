// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountCreator should render 1`] = `
<Wrapper
  onSubmit={[Function]}
/>
`;
